### ISSUE TITLE
Missing highlights for NearSpansUnordered inside a disjunction

### ISSUE THREAD

**Issue by mjustice3** [flaxsearch/luwak#120]
This is a difficult issue that I'm not sure can be fixed but I'm hoping to get your input. In TestHighlightingMatcher there are 2 tests for near spans inside a disjunction. Both are set up so that the near subqueries don't match. However, if I modify them so that the near's do match, then I see different behavior for the ordered vs unordered versions. The ordered version behaves as I would expect but the unordered version is missing some hits. Just wondering if you were aware of this discrepancy? The legacy profiling system that I'm trying to replace gets these missing hits for the unordered case. Will submit a pull request with the modified tests.

**Comment by mjustice3** [comment]
PR #121. Don't expect you to merge it just wanted to illustrate what I'm talking about.

**Comment by mjustice3** [comment]
Specifically, the problem seems to be with this loop in SpanExtractor:

                while (spans.nextStartPosition() != Spans.NO_MORE_POSITIONS) {
                    if (spans.startPosition() != -1)
                        spans.collect(collector);
                }

NearSpansUnordered.nextStartPosition() breaks down because subSpanCellsToPositionQueue() never gets called.

**Comment by mjustice3** [comment]
Added another test to PR #121. This one triggers the assertion in NearSpansUnordered.nextStartPosition() 

java.lang.AssertionError
        at org.apache.lucene.search.spans.NearSpansUnordered.nextStartPosition(NearSpansUnordered.java:228)
        at uk.co.flax.luwak.util.XNearSpansOrdered.advancePosition(XNearSpansOrdered.java:114)
        at uk.co.flax.luwak.util.XNearSpansOrdered.stretchToOrder(XNearSpansOrdered.java:98)
        at uk.co.flax.luwak.util.XNearSpansOrdered.nextStartPosition(XNearSpansOrdered.java:75)
        at uk.co.flax.luwak.util.SpanExtractor.collect(SpanExtractor.java:80)

**Comment by mjustice3** [comment]
Updated the assertion triggering test. In OTL the query looks like:

```
TEST <And>
* <Or>
** "now"
** <And>
*** "is"
*** <Order><Near/10>
**** "the"
**** <Near/5>
***** "time men"
***** "all"
* "good"
```

The inner conjunction fails but overall the query should pass against "now is the time for all good men". I expect to see 2 hits: 'now' and 'good'. My legacy system also reports 'is' as a hit but I don't think that should be there. If I replace the assert in NearSpansUnordered.nextStartPosition() with

```
    //assert minPositionCell().startPosition() != NO_MORE_POSITIONS;
    if (minPositionCell().startPosition() == NO_MORE_POSITIONS) {
        oneExhaustedInCurrentDoc = true;
        return NO_MORE_POSITIONS;
    }
```

then I get the same results as the legacy system.

**Comment by mjustice3** [comment]
The two failing tests added in PR #121 plus other unreported issues are all resolved by adding this method to DisjunctionScorer

```
  public final Collection<ChildScorer> getMatchingChildren() {
      ArrayList<ChildScorer> children = new ArrayList<>();
      try {
          for (DisiWrapper w = getSubMatches(); w != null; w = w.next) {
              children.add(new ChildScorer(w.scorer, "SHOULD"));
          }
      } catch (IOException e) {
      }
      return children;
    }
```

and then adding a special case in SpanExtractor.getSpans()

```
        Collection<Scorer.ChildScorer> children = null;
        if (scorer instanceof DisjunctionScorer) {
            children = ((DisjunctionScorer)scorer).getMatchingChildren();
        } else {
            children = scorer.getChildren();
        }
```

**Comment by romseygeek** [comment]
Oh, nice catch Marvin.  I think pushing the DisjunctionScorer change back up into Lucene core will mean we can ditch the luwak-specific forks of NearSpanQuery et al as well.  I'll open a JIRA.

**Comment by romseygeek** [comment]
I opened https://issues.apache.org/jira/browse/LUCENE-7628

**Comment by mjustice3** [comment]
Thanks Alan. In my local lucene branch I've made DisjunctionScorer public so that getMatchingChildren is accessible from the SpanExtractor but I guess that's not an acceptable long term solution?

**Comment by mjustice3** [comment]
FWIW, after the DisjunctionScorer  change,  I did a run of 100K docs against 17K queries with and without XSpanNearQuery and see zero differences.

**Comment by romseygeek** [comment]
This ought to be fixed by https://issues.apache.org/jira/browse/LUCENE-7628, which will be released either with lucene 6.5 or 7.0, whichever happens next.

**Comment by romseygeek** [comment]
Fixed by 1.5 release
